README.md
# data-exporter-service [![Build Status](http://grisham:8111/app/rest/builds/buildType:%28id:Books_Platform_Mimir_DataExporterService_BuildTestPublish%29/statusIcon)](http://grisham:8111/viewType.html?buildTypeId=Books_Platform_Mimir_DataExporterService_BuildTestPublish&guest=1)

[ [Release Notes](http://jira.blinkbox.local/confluence/display/REL/Data+Exporter+Service+-+Release+Note) ]

Service that exports snapshots of shop data, for reporting purposes.

## Requirements

The data export service will will periodically update snapshots of shop data such as books, publishers, clubcards etc
to the reporting database.

A detailed list of data that is required for reporting is 
given at [this page on Confluence](https://tools.mobcastdev.com/confluence/display/AN/Reporting+DB+requirements+-+Shop+Data), along with suggested priorities.

## Developer install

To build the service, you need to have [sbt](acceptance-test/data-export-service-test.properties.example) version 0.13 installed.

To run the service in development, use the command:

```
$ sbt run
```

By default, the service starts scheduling data export jobs at the given intervals. But there's also a command
line argument for running a one-off export, "--now". This is useful for testing and performing one-off update 
of reporting data. So, you can run it via sbt like this:

```
$ sbt run "--now"
```

To build a deployable version of the service, use the command:

```
$ sbt assembly
```

This builds a single JAR file that can be deployed without any other dependencies or application servers.
This is built using the sbt plugin [sbt-assembly](https://github.com/sbt/sbt-assembly), see 
its GitHub page for details on how it works.

You also need to create a configuration file to specify how the service will connect to databases.
There is an example configuration file in the source directory src/main/resources/data-export-service.properties.example.
Edit this with the appropriate parameters for your environment, rename it so the file has a ".properties" extension only,
and put this file on the classpath that's specified when running the service.

If you want to work on the code in an IDE, you'll want to generate the project files for your preferred IDE.
For example, for Eclipse you would do the following steps:

First ensure that you have [the sbt-eclipse plugin](https://github.com/typesafehub/sbteclipse) installed. Typically, for the sbt 0.13 onwards, that involves adding the following line in a file ~/.sbt/0.13/plugins/plugins.sbt:

```
addSbtPlugin("com.typesafe.sbteclipse" % "sbteclipse-plugin" % "2.4.0")
```

Once you have configured the plugin, you can generate the Eclipse project files using this command:

```
$ sbt eclipse with-source=true
```

## MySQL database creation

To set up the database tables, run the script db/schema.sql 
against the desired database server. For example:

```
$ mysql -u <username> -p <password> reporting < db/schema.sql
```

## Running the service

To run the service, you need to provide it with configuration, typically as a properties file.
An example file is in [src/main/resources/data-exporter-service.properties.example](https://git.mobcastdev.com/Mimir/data-exporter-service/blob/master/src/main/resources/data-exporter-service.properties.example).
Edit this as necessary, then put it in the directory you'll be running the service from.  

To start the service, use a command like:

```
$ java -jar data-export-service-assembly-1.0.jar
```

Note that the current directory "." is included on the classpath of the built jar file, so
you can run the service by putting the properties file here.


CHANGELOG.md
# Data Exporter Service Change Log

## 3.0.4 ([#25](https://git.mobcastdev.com/Mimir/data-exporter-service/pull/25) 2014-12-15 14:08:15)

CP-2083: Perform updates in smaller transactions

### Improvements:

- Updated dependencies and Scala version, and updated code to match.
- Changed data exporting code to ensure transactions aren't very big. This sacrifices transactionality in favour of an easier load for the database replication. The table deletion and copying code has been rearranged to minimise any potential impact of a DB error part way through a data export job.


## 3.0.3 ([#24](https://git.mobcastdev.com/Mimir/data-exporter-service/pull/24) 2014-10-17 16:09:46)

Fix memory usage and too long contributor URLs.

### Bug fixes:

- CP-1944: Enable streaming of query results from input databases, to avoid using a lot of memory when copying large tables.
- CP-1974: Truncate long generated contributor URLs to avoid potential failure when writing to contributors table.


## 3.0.2 ([#23](https://git.mobcastdev.com/Mimir/data-exporter-service/pull/23) 2014-09-22 15:15:05)

CP-1795 Fixed leaking DB connections

### Bug fix:

- Stop leaking DB connections, which caused the connection pool to run out of connections after a number of jobs, causing the service to stop running further jobs.


## 3.0.1 ([#22](https://git.mobcastdev.com/Mimir/data-exporter-service/pull/22) 2014-08-20 07:57:34)

Fixed incorrect column data-type

PATCH

## 3.0.0 ([#21](https://git.mobcastdev.com/Mimir/data-exporter-service/pull/21) 2014-08-07 13:25:12)

Moved discount from publisher to book and amended tests accordingly

### Breaking Change

* Discount is now dynamic (per-book) rather than per-publisher.

## 2.0.3 ([#20](https://git.mobcastdev.com/Mimir/data-exporter-service/pull/20) 2014-07-25 11:08:45)

Don't require DB on startup

### Improvements:

- Don't exit on startup if DB is not immediately available.


## 2.0.2 ([#19](https://git.mobcastdev.com/Mimir/data-exporter-service/pull/19) 2014-07-24 12:40:10)

Simplified config

### Improvements:

* Moved parts of config into reference.conf


## 1.0.4 ([#18](https://git.mobcastdev.com/Mimir/data-exporter-service/pull/18) 2014-07-23 17:04:58)

Bump minor version number.

Bug fix: bump minor version number.

## 1.0.3 (2014-06-05 12:00)

Last manually versioned build.


src/test/scala/com/blinkboxbooks/mimir/export/DbUtilsTest.scala
package com.blinkboxbooks.mimir.export

import org.scalatest.FunSuite
import org.junit.runner.RunWith
import org.scalatest.junit.JUnitRunner
import org.mockito.Mockito._
import org.mockito.Matchers._
import javax.sql.DataSource
import org.scalatest.mock.MockitoSugar
import org.squeryl.Session
import java.sql.Connection
import java.sql.SQLException
import org.scalatest.FlatSpec

@RunWith(classOf[JUnitRunner])
class DbUtilsTest extends FlatSpec with MockitoSugar {

  import DbUtils._

  trait ConnectionFixture {
    val datasource = mock[DataSource]
    val connection = mock[Connection]
    doReturn(connection).when(datasource).getConnection
    val fn = mock[(Session) => Unit]
  }

  "Managed transaction" should "commit changes after successful action" in new ConnectionFixture {
    DbUtils.withSession(datasource)(fn)

    verify(fn).apply(any[Session])
    verify(connection).setAutoCommit(false)
    verify(connection).commit()
    verify(connection).close()
    verifyNoMoreInteractions(fn, connection)
  }

  "Managed transaction" should "roll back connection when action fails" in new ConnectionFixture {
    val ex = new RuntimeException("Test exception")
    doThrow(ex).when(fn).apply(any[Session])

    val thrown = intercept[Exception] { DbUtils.withSession(datasource)(fn) }

    assert(thrown eq ex)
    verify(fn).apply(any[Session])
    verify(connection).setAutoCommit(false)
    verify(connection).rollback()
    verify(connection).close()
    verifyNoMoreInteractions(fn, connection)
  }

  "Managed transaction" should "roll back connection when commit fails" in new ConnectionFixture {
    val ex = new SQLException("Test exception")
    doThrow(ex).when(connection).commit()

    val thrown = intercept[Exception] { DbUtils.withSession(datasource)(fn) }
    assert(thrown eq ex)

    verify(fn).apply(any[Session])
    verify(connection).setAutoCommit(false)
    verify(connection).commit()
    verify(connection).rollback()
    verify(connection).close()
    verifyNoMoreInteractions(fn, connection)
  }

  "Read only connection" should "be closed after successful action" in new ConnectionFixture {
    DbUtils.withReadOnlySession(datasource, None)(fn)

    verify(fn).apply(any[Session])
    verify(connection).setAutoCommit(false)
    verify(connection).close()
    verifyNoMoreInteractions(fn, connection)
  }

  "Read only connection" should "be closed after failed action" in new ConnectionFixture {
    val ex = new RuntimeException("Test exception")
    doThrow(ex).when(fn).apply(any[Session])

    val thrown = intercept[Exception] { DbUtils.withReadOnlySession(datasource, Some(100))(fn) }
    assert(thrown eq ex)

    verify(fn).apply(any[Session])
    verify(connection).setAutoCommit(false)
    verify(connection).close()
    verifyNoMoreInteractions(fn, connection)
  }

}


src/test/scala/com/blinkboxbooks/mimir/export/HelperMethodsTest.scala
package com.blinkboxbooks.mimir.export

import org.scalatest.{BeforeAndAfter, BeforeAndAfterAll, FunSuite}
import org.junit.runner.RunWith
import org.scalatest.junit.JUnitRunner

@RunWith(classOf[JUnitRunner])
class BookMediaTest extends FunSuite {
  test("Return full jpg URL from base resource server URL") {
    val expectedConversions = Map(
      "https://media.blinkboxboks.com/path/to/image.png" -> "https://media.blinkboxboks.com/params;v=0/path/to/image.png.jpg",
      "https://media.blinkboxboks.com/path/to/image.jpg" -> "https://media.blinkboxboks.com/path/to/image.jpg"
    )
    val converted = expectedConversions.keySet.map(baseUrl => BookMedia.fullsizeJpgUrl(Some(baseUrl))).flatten
    assert(converted === expectedConversions.values.toSet)
  }

  test("Return None on non-URLs being converted to full jpg URL") {
    val converted = BookMedia.fullsizeJpgUrl(Some("Not a URL"))
    assert(converted === None)
  }
}

@RunWith(classOf[JUnitRunner])
class ContributorTest extends FunSuite {
  val baseUrl = "https://www.blinkboxbooks.com/#!/author"

  test("Generates author URLs") {
    val actualUrl = Contributor.generateContributorUrl(baseUrl,"guid","F. Honda Dïcks")
    val expectedUrl = Some("https://www.blinkboxbooks.com/#!/author/guid/f-honda-dicks")
    assert(actualUrl === expectedUrl)
  }

  test("Deals with empty name") {
    val actualUrl = Contributor.generateContributorUrl(baseUrl, "guid","")
    val expectedUrl = Some("https://www.blinkboxbooks.com/#!/author/guid/details")
    assert(actualUrl === expectedUrl)
  }
  
  test("Deals with a very long author name") {
    val longAuthorName = "abcdefghij" * 30
    val actualUrl = Contributor.generateContributorUrl(baseUrl, "guid", longAuthorName)
    assert(actualUrl.get.size == 255, s"Actual URL: $actualUrl, author name: $longAuthorName")
    assert(actualUrl.get.startsWith("https://www.blinkboxbooks.com/#!/author/guid/abcdefghijabcdefghij"))
  }

}


src/main/scala/com/blinkboxbooks/mimir/export/DbUtils.scala
package com.blinkboxbooks.mimir.export

import com.typesafe.config.Config
import java.sql.Connection
import java.sql.PreparedStatement
import java.sql.ResultSet
import java.sql.Statement
import javax.sql.DataSource
import org.apache.commons.dbcp.BasicDataSource
import org.squeryl.Session
import org.squeryl.adapters.MySQLAdapter
import scala.util.control.NonFatal

object DbUtils {

  /**
   * Create a datasource for the given MySQL DB.
   */
  def createDatasource(config: Config) = {
    val datasource = new BasicDataSource
    datasource.setUrl(config.getString("url"))
    datasource.setUsername(config.getString("username"))
    datasource.setPassword(config.getString("password"))
    datasource.setDriverClassName(config.getString("driver"))
    datasource.setValidationQuery("SELECT 1")
    datasource
  }

  /**
   * Run the given action, providing it a session for the database, and treating any
   * changes as a single transaction. On error, any changes will be rolled back.
   */
  def withSession(datasource: DataSource)(fn: (Session) => Unit) = {
    val connection = datasource.getConnection()
    try {
      connection.setAutoCommit(false)
      val session: Session = Session.create(connection, new StreamingMySqlAdapter(None))
      fn(session)
      connection.commit()
    } catch {
      case NonFatal(e) => {
        connection.rollback()
        throw e
      }
    } finally {
      connection.close()
    }
  }

  /**
   * Run the given action, providing it a session for the database.
   */
  def withReadOnlySession(datasource: DataSource, fetchSize: Option[Int])(fn: (Session) => Unit) = {
    val connection = datasource.getConnection()
    try {
      connection.setAutoCommit(false)
      val session: Session = Session.create(connection, new StreamingMySqlAdapter(fetchSize))
      fn(session)
    } finally {
      connection.close()
    }
  }

  class StreamingMySqlAdapter(fetchSize: Option[Int]) extends MySQLAdapter {

    override def prepareStatement(conn: Connection, statement: String): PreparedStatement = {
      val stmt = conn.prepareStatement(statement, ResultSet.TYPE_FORWARD_ONLY, ResultSet.CONCUR_READ_ONLY)
      fetchSize.foreach(s => stmt.setFetchSize(s))
      stmt
    }

    override def createStatement(conn: Connection): Statement = {
      val stmt = conn.createStatement(ResultSet.TYPE_FORWARD_ONLY, ResultSet.CONCUR_READ_ONLY)
      fetchSize.foreach(s => stmt.setFetchSize(s))
      stmt
    }

  }

}

src/main/scala/com/blinkboxbooks/mimir/export/StreamTest.scala
package com.blinkboxbooks.mimir.export

import org.apache.commons.dbcp.BasicDataSource
import com.typesafe.scalalogging.slf4j.Logging
import org.squeryl.{ SessionFactory, Session, Schema }
import org.squeryl.adapters.MySQLAdapter
import org.squeryl.PrimitiveTypeMode._
import rx.lang.scala.Observable
import scala.concurrent.{ Await, promise }
import scala.concurrent.duration._

// Input schemas.
case class Publisher(id: Int, name: String, ebookDiscount: Int,
  implementsAgencyPricingModel: Boolean, countryCode: Option[String]) {
  def this() = this(-1, "", -1, false, None)
}

// Output schemas.
case class PublisherInfo(id: Int, name: String, ebookDiscount: Int,
  implementsAgencyPricingModel: Boolean, countryCode: Option[String]) {
  def this() = this(-1, "", -1, false, None)
}

object StreamTest extends App with Schema with Logging {

  val bufferSize = 5

  // Connect to a DB.
  logger.info("Starting")

  // Input schema defs.
  val publisherData = table[Publisher]("dat_publisher")
  on(publisherData)(p => declare(
    p.ebookDiscount is (named("ebook_discount")),
    p.implementsAgencyPricingModel is (named("implements_agency_pricing_model")),
    p.ebookDiscount is (named("ebook_discount")),
    p.countryCode is (named("country_code"), dbType("VARCHAR(4)"))))

  // Output schema defs.
  val publishersOutput = table[PublisherInfo]("publishers")
  on(publishersOutput)(p => declare(
    p.ebookDiscount is (named("ebook_discount")),
    p.implementsAgencyPricingModel is (named("implements_agency_pricing_model")),
    p.ebookDiscount is (named("ebook_discount")),
    p.countryCode is (named("country_code"), dbType("VARCHAR(4)"))))

  // Configure reporting database that we'll be writing to.
  val outputDatasource = new BasicDataSource
  outputDatasource.setDriverClassName("com.mysql.jdbc.Driver")
  outputDatasource.setUrl("jdbc:mysql://localhost/reporting")
  outputDatasource.setUsername("gospoken")
  outputDatasource.setPassword("gospoken")
  configureThreadPoolParameters(outputDatasource)
  outputDatasource.getConnection.close

  // Configure shop database that we'll get data from.
  val shopDatasource = new BasicDataSource
  shopDatasource.setDriverClassName("com.mysql.jdbc.Driver")
  shopDatasource.setUrl("jdbc:mysql://localhost/shop")
  shopDatasource.setUsername("gospoken")
  shopDatasource.setPassword("gospoken")
  configureThreadPoolParameters(shopDatasource)
  shopDatasource.getConnection.close

  // The global singleton session factory (yuck) refers to the shop database.
  SessionFactory.concreteFactory =
    Some(() => Session.create(shopDatasource.getConnection(), new MySQLAdapter))
  SessionFactory.newSession.bindToCurrentThread

  //printDdl(str => println(str))

  val connection = outputDatasource.getConnection()
  try {
    val outputSession: Session = Session.create(connection, new MySQLAdapter)
    connection.setAutoCommit(false)

    // Clear old data.
    using(outputSession) { publishersOutput.deleteWhere(p => p.id gte 0) }

    //throw new RuntimeException("TESTING TESTING")

    val results = from(publisherData)(p => where(p.id gte 0) select (p))

    val prom = promise[Unit]
    val o = Observable.from(results)
      .map(pub => new PublisherInfo(pub.id, pub.name, pub.ebookDiscount, pub.implementsAgencyPricingModel, pub.countryCode))
      .buffer(bufferSize)
      .doOnEach(pubs => using(outputSession) { publishersOutput.insert(pubs) })
      .doOnError(e => { prom.failure(e); e.printStackTrace(System.err) })
      .doOnCompleted(() => { prom.success() })

    logger.info("Executing copy")

    o.subscribe

    // Wait, with timeout.
    Await.result(prom.future, 1 minute)

    connection.commit()
  } catch {
    case t: Throwable => {
      logger.error("Error while executing database job", t)
      connection.rollback()
    }
  } finally {
    connection.close()
  }

  logger.info("Completed")

  private def configureThreadPoolParameters(datasource: BasicDataSource) = {
    datasource.setMaxActive(10)
    datasource.setMaxIdle(5)
    datasource.setInitialSize(5)
    datasource.setValidationQuery("SELECT 1")
  }

}


src/main/scala/com/blinkboxbooks/mimir/export/DataExporterService.scala
package com.blinkboxbooks.mimir.export

import com.blinkbox.books.logging.Loggers
import com.blinkbox.books.config.Configuration
import com.typesafe.scalalogging.StrictLogging
import it.sauronsoftware.cron4j.Scheduler
import java.sql.Date
import java.util.concurrent.atomic.AtomicLong
import java.util.concurrent.TimeUnit
import javax.sql.DataSource
import org.apache.commons.dbcp.BasicDataSource
import org.squeryl.{ SessionFactory, Session, Schema, Query, Table }
import org.squeryl.adapters.MySQLAdapter
import org.squeryl.PrimitiveTypeMode._
import rx.lang.scala.Observable
import scala.concurrent.{ Await, promise }
import scala.concurrent.duration._
import scala.concurrent.Promise

object DataExporterService extends App with Configuration with StrictLogging with Loggers {

  import ShopSchema._
  import ClubcardSchema._
  import ReportingSchema._
  import DbUtils._

  val serviceConfig = config.getConfig("service.dataExporter")

  val batchSize = serviceConfig.getInt("jdbcBatchsize")
  val timeout = serviceConfig.getDuration("jdbcTimeout", TimeUnit.MILLISECONDS).millis
  val fetchSize = serviceConfig.getInt("fetchSize")
  val authorBaseUrl = serviceConfig.getString("authorBaseUrl")

  // Configure datasources for reading and writing.
  val shopDatasource = createDatasource(serviceConfig.getConfig("shopDb"))
  val clubcardDatasource = createDatasource(serviceConfig.getConfig("clubcardDb"))
  val outputDatasource = createDatasource(serviceConfig.getConfig("reportingDb"))

  if (args.contains("--now")) {
    logger.info("Starting one-off export")
    runDataExport(shopDatasource, clubcardDatasource, outputDatasource, batchSize, timeout, fetchSize)
    logger.info("Completed export")
  } else {
    val cronStr = serviceConfig.getString("schedule")
    logger.info(s"Scheduling data export with configuration: $cronStr")
    val scheduler = new Scheduler()
    scheduler.schedule(cronStr, new Runnable() {
      override def run(): Unit = {
        logger.info("Starting scheduled export")
        try {
          runDataExport(shopDatasource, clubcardDatasource, outputDatasource, batchSize, timeout, fetchSize)
          logger.info("Completed scheduled export")
        } catch {
          case e: Exception => logger.error("Failed data export", e)
        }
      }
    })
    scheduler.setDaemon(false)
    scheduler.start()
  }

  /**
   * Perform all the data export jobs.
   */
  def runDataExport(shopDatasource: DataSource, clubcardDatasource: DataSource, outputDatasource: DataSource,
                    batchSize: Int, timeout: Duration, fetchSize: Int, authorBaseUrl: String = authorBaseUrl): Unit = {

    implicit val t = timeout
    implicit val b = batchSize

    withSession(outputDatasource)(implicit outputSession => {

      // Write new snapshots of all tables.
      withReadOnlySession(shopDatasource, Some(fetchSize))(shopSession => {
        using(shopSession) {

          syncTable(from(publisherData)(select(_)), publishersOutput, identity[Publisher])
          syncTable(from(mapBookContributorData)(select(_)), contributorRolesOutput, identity[MapBookToContributor])
          syncTable(from(genreData)(select(_)), genresOutput, identity[Genre])
          syncTable(from(bookGenreData)(select(_)), bookGenresOutput, identity[MapBookToGenre])
          syncTable(from(currencyRateData)(select(_)), currencyRatesOutput, identity[CurrencyRate])

          val contributorConverter = (c: Contributor) =>
            new OutputContributor(c.id, c.fullName, c.firstName, c.lastName, c.guid, BookMedia.fullsizeJpgUrl(c.imageUrl), Contributor.generateContributorUrl(authorBaseUrl, c.guid, c.fullName))
          syncTable(from(contributorData)(select(_)), contributorsOutput, contributorConverter)

          val bookResults =
            join(bookData, bookMediaData.leftOuter)((book, media) =>
              where(media.map(_.kind) === BookMedia.BOOK_COVER_MEDIA_ID or (media.map(_.kind).isNull))
                select (book, media.getOrElse(new BookMedia(1, book.id, Some(""), BookMedia.BOOK_COVER_MEDIA_ID)))
                on (book.id === media.map(_.isbn).get))
          val bookConverter = (b: (Book, BookMedia)) =>
            new OutputBook(b._1.id, b._1.publisherId, b._1.discount, b._1.publicationDate, b._1.title, b._1.description.map({ _.take(ReportingSchema.MAX_DESCRIPTION_LENGTH) }),
              b._1.languageCode, b._1.numberOfSections, BookMedia.fullsizeJpgUrl(b._2.url))
          syncTable(bookResults, booksOutput, bookConverter, wait = true)
        }
      })

      withReadOnlySession(clubcardDatasource, Some(fetchSize))(clubcardSession => {
        using(clubcardSession) {
          val clubcardResults =
            from(clubcards, users, clubcardUsers)((clubcard, user, link) =>
              where(clubcard.id === link.cardId and user.id === link.userId)
                select (clubcard, user))
          val converter = (cu: (Clubcard, ClubcardUser)) =>
            new UserClubcardInfo(cu._1.cardNumber, Integer.parseInt(cu._2.userId))
          syncTable(clubcardResults, userClubcardsOutput, converter)
        }
      })
    })
  }

  /**
   * Delete the content of the given output table and syncTable it with data from the input.
   */
  def syncTable[T1, T2](input: Iterable[T1], output: Table[T2], converter: T1 => T2, wait: Boolean = false)(
    implicit bufferSize: Int, outputSession: Session, timeout: Duration): Unit = {
    clearTable(output)
    copy(input, output, converter)
  }

  /** Delete the content of the given table. */
  def clearTable[T](table: Table[T])(implicit session: Session) = using(session) { table.deleteWhere(r => 1 === 1) }

  /**
   * Copy the results from the given Query to the given output table, converting objects
   * using the given converter.
   *
   * Objects from the input query are streamed and written to the output database using
   * batched writes, for performance. The overall copy job is synchronous, in that it will wait for
   * the copying to complete before returning.
   */
  def copy[T1, T2](input: Iterable[T1], output: Table[T2], converter: T1 => T2, wait: Boolean = false)(
    implicit bufferSize: Int, outputSession: Session, timeout: Duration): Unit = {

    logger.info(s"Executing data export to table ${output.name}")

    val count = new AtomicLong(0)
    val p = Promise[Unit]()

    val observable = Observable.from(input)
      .map(converter)
      .buffer(bufferSize)

    // CP-2083: explicitly committing each batch, to avoid very big transactions.
    observable.subscribe(
      entities => using(outputSession) { output.insert(entities); outputSession.connection.commit(); count.addAndGet(entities.size) },
      e => { p.failure(e) },
      () => p.success(()))

    Await.result(p.future, timeout)
    logger.info(s"Completed export of ${count.get} rows")
  }

  /**
   *  Limit description field as that's potentially bigger in the incoming database - the Shop DB version allows
   *  longer varchars than the reporting database allows.
   */
  def truncate(book: Book): Book = book.copy(description = book.description.map(_.take(ReportingSchema.MAX_DESCRIPTION_LENGTH)))

}


src/main/scala/com/blinkboxbooks/mimir/export/Schemas.scala
package com.blinkboxbooks.mimir.export

import java.sql.Date
import org.squeryl.Schema
import org.squeryl.PrimitiveTypeMode._

// 
// Objects for schemas.
//
case class Book(id: String, publisherId: String, discount: Float, publicationDate: Date,
  title: String, description: Option[String], languageCode: Option[String], numberOfSections: Int) {
  def this() = this("", "", 0, new Date(0), "", None, None, 0)
}
case class Publisher(id: Int, name: String, 
  implementsAgencyPricingModel: Boolean, countryCode: Option[String]) {
  def this() = this(0, "", false, None)
}
case class Genre(id: Int, parentId: Option[Int], bisacCode: Option[String], name: Option[String]) {
  def this() = this(0, None, None, None)
}
case class MapBookToGenre(isbn: String, genreId: Int) {
  def this() = this("", 0)
}
case class CurrencyRate(fromCurrency: String, toCurrency: String, rate: BigDecimal) {
  def this() = this("", "", 0)
}
case class Contributor(id: Int, fullName: String, firstName: Option[String], lastName: Option[String], guid: String, imageUrl: Option[String]) {
  def this() = this(0, "", None, None, "", None)
}
case class MapBookToContributor(contributorId: Int, isbn: String, role: Int) {
  def this() = this(0, "", 0)
}
case class BookMedia(id: Int, isbn: String, url: Option[String], kind: Int){
  def this() = this(0, "", Some(""), 0)
}

// Enriched Output Classes
// They contain additional fields that are not in the source data.

case class OutputBook(id: String, publisherId: String, discount: Float, publicationDate: Date, title: String, description: Option[String],
                         languageCode: Option[String], numberOfSections: Int, coverUrl: Option[String]) {
  def this() = this("", "", 0, new Date(0), "", None, None, 0, None)
}
case class OutputContributor(id: Int, fullName: String, firstName: Option[String], lastName: Option[String], guid: String, imageUrl: Option[String], url: Option[String]) {
  def this() = this(0, "", None, None, "", None, None)
}


// Database enum values
object BookMedia {
  import java.net.URL
  val BOOK_COVER_MEDIA_ID = 0
  val FULL_EPUB_MEDIA_ID = 1
  val SAMPLE_EPUB_MEDIA_ID = 2

  def fullsizeJpgUrl(mediaUrl: Option[String]):Option[String] = mediaUrl.map { url =>
    try { new URL(url) } catch {
      case ex: Exception =>
        return None
    }
    url.takeRight(4) match {
      case ".jpg" => url
      case _ => url.replaceFirst("([^/])/([^/])", "$1/params;v=0/$2") + ".jpg"
    }
  }
}

object Contributor {
  import java.text.Normalizer

  def generateContributorUrl(baseUrl: String, guid: String, fullName: String): Option[String] = {
    val normalizedName = Normalizer.normalize(fullName, java.text.Normalizer.Form.NFD)
      .toLowerCase()
      .replaceAll(" ", "-")
      .replaceAll("[^a-z-]+", "")
    val normalizedNameWithDefault = if (normalizedName.isEmpty) "details" else normalizedName
    Some(Array(baseUrl, guid, normalizedNameWithDefault).mkString("/").take(255))
  }
}

// 
// Input schema definitions. 
//
// These are not full definitions of the corresponding tables, they just specify the columns
// that we read and the info needed to read them (e.g. column name but not details about column types).
//

object ShopSchema extends Schema {

  val bookData = table[Book]("dat_book")
  on(bookData)(b => declare(
    b.id is (named("isbn")),
    b.publisherId is (named("publisher_id")),
    b.discount is (named("discount")),
    b.publicationDate is (named("publication_date")),
    // Shop DB has longer size of this field than the reporting DB, hence replicate this in tests.
    b.description is dbType(s"varchar(65535)"),
    b.languageCode is (named("language_code")),
    b.numberOfSections is (named("num_sections"))))

  val publisherData = table[Publisher]("dat_publisher")
  on(publisherData)(p => declare(
    p.implementsAgencyPricingModel is (named("implements_agency_pricing_model")),
    p.countryCode is (named("country_code"))))

  val contributorData = table[Contributor]("dat_contributor")
  on(contributorData)(c => declare(
    c.guid is (named("guid")),
    c.fullName is (named("full_name")),
    c.imageUrl is (named("photo")),
    c.firstName is (named("first_name")),
    c.lastName is (named("last_name"))))

  val mapBookContributorData = table[MapBookToContributor]("map_book_contributor")
  on(mapBookContributorData)(m => declare(
    m.contributorId is (named("contributor_id"))))

  val genreData = table[Genre]("dat_genre")
  on(genreData)(g => declare(
    g.parentId is (named("parent_id")),
    g.bisacCode is (named("bisac_code"))))

  val bookGenreData = table[MapBookToGenre]("map_book_genre")
  on(bookGenreData)(g => declare(
    g.genreId is (named("genre_id"))))

  val currencyRateData = table[CurrencyRate]("dat_currency_rate")
  on(currencyRateData)(e => declare(
    e.fromCurrency is (named("from_currency")),
    e.toCurrency is (named("to_currency"))))

  val bookMediaData = table[BookMedia]("dat_book_media")
  on(bookMediaData)(m => declare(
    m.id is named("id"),
    m.isbn is named("isbn"),
    m.url is named("url"),
    m.kind is named("type")
  ))

}

case class Clubcard(id: Int, cardNumber: String) {
  def this() = this(0, "")
}
case class ClubcardUser(id: Int, userId: String) {
  def this() = this(0, "")
}
case class ClubcardForUser(cardId: Int, userId: Int) {
  def this() = this(0, 0)
}

object ClubcardSchema extends Schema {

  val clubcards = table[Clubcard]("club_card")
  on(clubcards)(c => declare(
    c.cardNumber is (named("card_number"))))

  val users = table[ClubcardUser]("users")
  on(users)(u => declare(
    u.userId is (named("user_id"))))

  val clubcardUsers = table[ClubcardForUser]("club_card_users")
  on(clubcardUsers)(cu => declare(
    cu.cardId is (named("card_id")),
    cu.userId is (named("user_id"))))

}

//
// Output schema definitions.
//

case class UserClubcardInfo(cardId: String, userId: Int) {
  def this() = this("", 0)
}

object ReportingSchema extends Schema {

  val MAX_DESCRIPTION_LENGTH = 20000

  val booksOutput = table[OutputBook]("books")
  on(booksOutput)(b => declare(
    b.id is (named("isbn")),
    b.publisherId is (named("publisher_id")),
    b.discount is (named("discount")),
    b.publicationDate is (named("publication_date"), dbType("DATE")),
    b.title is dbType("VARCHAR(255)"),
    b.description is dbType(s"varchar($MAX_DESCRIPTION_LENGTH)"),
    b.languageCode is (named("language_code"), dbType("CHAR(2)")),
    b.numberOfSections is (named("number_of_sections")),
    b.coverUrl is named("cover_url")))

  val publishersOutput = table[Publisher]("publishers")
  on(publishersOutput)(p => declare(
    p.implementsAgencyPricingModel is (named("implements_agency_pricing_model")),
    p.name is (dbType("VARCHAR(128)")),
    p.countryCode is (named("country_code"), dbType("VARCHAR(4)"))))

  val userClubcardsOutput = table[UserClubcardInfo]("user_clubcards")
  on(userClubcardsOutput)(c => declare(
    c.cardId is (named("clubcard_id"), dbType("VARCHAR(20)")),
    c.userId is (named("user_id"))))

  val currencyRatesOutput = table[CurrencyRate]("currency_rates")
  on(currencyRatesOutput)(e => declare(
    e.fromCurrency is (named("from_currency"), dbType("VARCHAR(5)")),
    e.toCurrency is (named("to_currency"), dbType("VARCHAR(5)"))))

  val contributorsOutput = table[OutputContributor]("contributors")
  on(contributorsOutput)(c => declare(
    c.fullName is (named("full_name"), dbType("VARCHAR(256)")),
    c.firstName is (named("first_name"), dbType("VARCHAR(256)")),
    c.lastName is (named("last_name"), dbType("VARCHAR(256)")),
    c.guid is (named("guid"), dbType("VARCHAR(256)")),
    c.url is (named("url"), dbType("VARCHAR(256)")),
    c.imageUrl is (named("image_url"), dbType("VARCHAR(256)"))))

  val contributorRolesOutput = table[MapBookToContributor]("contributor_roles")
  on(contributorRolesOutput)(m => declare(
    m.contributorId is (named("contributor_id"))))

  val genresOutput = table[Genre]("genres")
  on(genresOutput)(g => declare(
    g.parentId is (named("parent_id")),
    g.bisacCode is (named("bisac_code"), dbType("VARCHAR(8)"))))

  val bookGenresOutput = table[MapBookToGenre]("book_genres")
  on(bookGenresOutput)(g => declare(
    g.genreId is (named("genre_id")),
    g.isbn is (dbType("VARCHAR(13)"))))

  //  printDdl(str => println(str))

}
